### PR TITLE
resources: host detection + memory budgeting (issue #152)

### PR DESCRIPTION
**Pull request by grzanka**
This pull request introduces a memory budgeting feature to the `osh` application, allowing users to estimate and control the memory usage of scoring accumulators before running a simulation. It adds resource introspection, budget parsing, and enforcement of memory limits, as well as new CI steps to report host resources. The changes are grouped below by theme.

**Memory Budgeting and Enforcement:**

* Added a new `osh_membudget` module (`osh_membudget.c`, `osh_membudget.h`) that parses user-supplied memory budgets (plain bytes, unit-suffixed, or percentage forms) and computes a default budget based on host resources, with policies to ensure system stability. [[1]](diffhunk://#diff-3591be7a5070a5c1894dffabbecebdb05b24784ea680a840e96426fb59fb85acR1-R228) [[2]](diffhunk://#diff-97e04cb3ba7f07990509cb5b1a334cf2579d95a0a58aa385cedcce716bd2a3acR1-R69)
* Integrated memory budget checks into the scoring workflow: before allocating scoring accumulators, the code now estimates memory needs and refuses to proceed if the run would exceed the user-specified or default budget. [[1]](diffhunk://#diff-559d5132c2ba04439af7100acff3d22eca02e07f1c92ccf0b3557bdf878bdd91R48-R49) [[2]](diffhunk://#diff-559d5132c2ba04439af7100acff3d22eca02e07f1c92ccf0b3557bdf878bdd91R270-R277) [[3]](diffhunk://#diff-f4ef3e337a8b21ebfd493b60e131d559b38c7b2b783dfcc8a77167b91a939b1fR95)
* Updated the build system to include the new `osh_membudget.c` source file in the application build.

**Host Resource Introspection:**

* Added a new CLI action (`--print-resources`) to print detected host CPU and RAM resources, using the new `osh_sysinfo` functionality. This is now also exercised in CI workflows for all platforms. [[1]](diffhunk://#diff-f4ef3e337a8b21ebfd493b60e131d559b38c7b2b783dfcc8a77167b91a939b1fR8) [[2]](diffhunk://#diff-f4ef3e337a8b21ebfd493b60e131d559b38c7b2b783dfcc8a77167b91a939b1fR47-R66) [[3]](diffhunk://#diff-faff1af3d8ff408964a57b2e475f69a6b7c7b71c9978cccc8f471798caac2c88R63-R71)

**API and Header Updates:**

* Extended the scoring API with a memory estimate structure and function (`osh_scoring_mem_estimate`, `osh_scoring_estimate_memory`) to predict the memory footprint of scoring accumulators before allocation. [[1]](diffhunk://#diff-e8aa84da844a4e5d7724b3473e988dfc03360b64eab42068f7ac1d82636b0fd6R5) [[2]](diffhunk://#diff-e8aa84da844a4e5d7724b3473e988dfc03360b64eab42068f7ac1d82636b0fd6R68-R121)

### CODE REVIEW

**Comment by grzanka** [comment]
@copilot remove completely the "GPU detection" line from the printout

File: src/apps/osh/main.c [hunk 1]
```diff
@@ -43,6 +44,27 @@ int main(int argc, char *argv[]) {
...
+
+        printf("Host resources:\n");
+        if (info.logical_cores > 0u) {
+            printf("  Logical CPU cores : %u\n", info.logical_cores);
+        } else {
+            printf("  Logical CPU cores : unknown\n");
+        }
+        printf("  Total RAM         : %s\n", info.ram_total_bytes > 0u ? total : "unknown");
+        printf("  Available RAM     : %s\n", info.ram_available_bytes > 0u ? avail : "unknown");
+        printf("  GPU detection     : not yet implemented\n");
```
Comment (grzanka) [inline src/apps/osh/main.c]:
```suggestion
```

**Comment by grzanka** [comment]
@copilot fix failing tests

**Comment by grzanka** [comment]
Original PR description for first part was:

Implements **issue #152** (parallelization *preparation*, not optimization). Per the plan, this PR is being built in two components on the **same branch**:

- ✅ **Component A — host resource detection** (this push)
- ⏳ **Component B — scoring memory estimate + budget + OOM gate** (to be added to this same PR after review of A)

Opened as a **draft** so Component A can be reviewed before B lands here.

---

## Component A — what's included

A best-effort host-resource probe, isolated behind one platform module so the rest of the tree stays portable and the **core library never calls the OS** (detection is an app/embedder concern).

**`src/common/osh_sysinfo.{h,c}`** — the project's single place for OS resource queries (the resource counterpart of `osh_time.c`):

```c
struct osh_sysinfo { unsigned logical_cores; uint64_t ram_total_bytes;
                     uint64_t ram_available_bytes; unsigned gpu_count; };
enum osh_status osh_sysinfo_query(struct osh_sysinfo *out);   /* best-effort, never hard-fails */
void osh_sysinfo_format_bytes(uint64_t bytes, char *buf, size_t buflen);  /* "61.2 GiB" */
```

Per-platform backends, all behind `#ifdef`:

| Platform | cores | total | available |
|---|---|---|---|
| Linux | `sysconf(_SC_NPROCESSORS_ONLN)` | physical **∧ cgroup limit** (min) | `/proc/meminfo` `MemAvailable` |
| macOS | `hw.logicalcpu` | `hw.memsize` | `host_statistics64` (free+inactive) |
| Windows | `GetActiveProcessorCount` | `GlobalMemoryStatusEx.ullTotalPhys` | `…ullAvailPhys` |
| WASM | `emscripten_num_logical_cores` | `emscripten_get_heap_max` | heap max − heap size |
| other | zero-fill | | |

Unknown fields stay `0`; the cgroup handling addresses the classic "ignore the container limit → OOMKilled" trap. `gpu_count` is scaffolded (always 0) for the later GPU work. `osh_sysinfo_format_bytes` is here because Component B's memory report reuses it.

**CLI** — new `--print-resources` action:
```
$ openshieldhit --print-resources
Host resources:
  Logical CPU cores : 4
  Total RAM         : 15.7 GiB
  Available RAM     : 15.0 GiB
  GPU detection     : not yet implemented
```

## Why it's safe / preparation-only
- Adds **no parallelism** and touches **no hot path** — detection runs only when asked.
- Core transport/scoring is unchanged; `osh_sysinfo` is a leaf utility in `common/` that core code does not call.
- Designed to extend cleanly into B: the struct already carries exactly what B's budget policy needs, and the byte formatter is shared.

## Tests (run on every CI OS: ubuntu / windows / macos)
- `tests/unit/test_osh_sysinfo.c`: `query` invariants (`cores ≥ 1`, `total > 0`, `available ≤ total`, `gpu_count` reset), `NULL` guard, and the formatter including the tiny-buffer / `buflen==0` edge cases.
- `unit::cli::print_resources`: end-to-end smoke test of the executable, gated on the `Total RAM` line.

## Verification
- Clean build, no warnings; `clang-format-19 --dry-run --Werror` clean on the whole tree.
- Full suite green except the 4 pre-existing environmental failures (missing LFS DICOM fixtures + git-version-string), unrelated to this change.
- I can only run Linux locally; the Windows/macOS/WASM branches are written against their documented APIs and rely on the CI matrix for compile/run verification — please watch those jobs on first CI run.

## Not in this PR (coming to this same branch as Component B)
Scoring memory estimator (`Σ pages len·8·k`), budget policy (80%-of-available, cgroup/WASM-aware, `--mem-budget`), the setup-time memory report, the hard OOM gate, `--threads`/`--score-reduction` flags, the profile-JSON `resources`/`memory` blocks, and the new cylindrical + large-mesh benchmark cases.

Part of #152.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
_Generated by [Claude Code](https://claude.ai/code/session_01WKLBq2r6GWPR3bKwa37mdm)_

**Review by nbassler — APPROVED** [review]
looks ok to me
